README

 libmad - MPEG audio decoder library
 Copyright (C) 2000-2004 Underbit Technologies, Inc.

 $Id: README,v 1.4 2004/01/23 09:41:32 rob Exp $

===============================================================================

INTRODUCTION

  MAD (libmad) is a high-quality MPEG audio decoder. It currently supports
  MPEG-1 and the MPEG-2 extension to Lower Sampling Frequencies, as well as
  the so-called MPEG 2.5 format. All three audio layers (Layer I, Layer II,
  and Layer III a.k.a. MP3) are fully implemented.

  MAD does not yet support MPEG-2 multichannel audio (although it should be
  backward compatible with such streams) nor does it currently support AAC.

  MAD has the following special features:

    - 24-bit PCM output
    - 100% fixed-point (integer) computation
    - completely new implementation based on the ISO/IEC standards
    - distributed under the terms of the GNU General Public License (GPL)

  Because MAD provides full 24-bit PCM output, applications using MAD are
  able to produce high quality audio. Even when the output device supports
  only 16-bit PCM, applications can use the extra resolution to increase the
  audible dynamic range through the use of dithering or noise shaping.

  Because MAD uses integer computation rather than floating point, it is
  well suited for architectures without a floating point unit. All
  calculations are performed with a 32-bit fixed-point integer
  representation.

  Because MAD is a new implementation of the ISO/IEC standards, it is
  unencumbered by the errors of other implementations. MAD is NOT a
  derivation of the ISO reference source or any other code. Considerable
  effort has been expended to ensure a correct implementation, even in cases
  where the standards are ambiguous or misleading.

  Because MAD is distributed under the terms of the GPL, its redistribution
  is not generally restricted, so long as the terms of the GPL are followed.
  This means MAD can be incorporated into other software as long as that
  software is also distributed under the GPL. (Should this be undesirable,
  alternate arrangements may be possible by contacting Underbit.)

===============================================================================

ABOUT THE CODE

  The code is optimized and performs very well, although specific
  improvements can still be made. The output from the decoder library
  consists of 32-bit signed linear fixed-point values that can be easily
  scaled for any size PCM output, up to 24 bits per sample.

  The API for libmad can be found in the `mad.h' header file. Note that this
  file is automatically generated, and will not exist until after you have
  built the library.

  There are two APIs available, one high-level, and the other low-level.
  With the low-level API, each step of the decoding process must be handled
  explicitly, offering the greatest amount of control. With the high-level
  API, after callbacks are configured, a single routine will decode an
  entire bitstream.

  The high-level API may either be used synchronously or asynchronously. If
  used asynchronously, decoding will occur in a separate process.
  Communication is possible with the decoding process by passing control
  messages.

  The file `minimad.c' contains an example usage of the libmad API that
  shows only the bare minimum required to implement a useful decoder. It
  expects a regular file to be redirected to standard input, and it sends
  decoded 16-bit signed little-endian PCM samples to standard output. If a
  decoding error occurs, it is reported to standard error and decoding
  continues. Note that the scale() routine in this code is only provided as
  an example; it rounds MAD's high-resolution samples down to 16 bits, but
  does not perform any dithering or noise shaping. It is therefore not
  recommended to use this routine as-is in your own code if sound quality is
  important.

Integer Performance

  To get the best possible performance, it is recommended that an assembly
  version of the fixed-point multiply and related routines be selected.
  Several such assembly routines have been written for various CPUs.

  If an assembly version is not available, a fast approximation version will
  be used. This will result in reduced accuracy of the decoder.

  Alternatively, if 64-bit integers are supported as a datatype by the
  compiler, another version can be used that is much more accurate.
  However, using an assembly version is generally much faster and just as
  accurate.

  More information can be gathered from the `fixed.h' header file.

  MAD's CPU-intensive subband synthesis routine can be further optimized at
  the expense of a slight loss in output accuracy due to a modified method
  for fixed-point multiplication with a small windowing constant. While this
  is helpful for performance and the output accuracy loss is generally
  undetectable, it is disabled by default and must be explicitly enabled.

  Under some architectures, other special optimizations may also be
  available.

Audio Quality

  The output from MAD has been found to satisfy the ISO/IEC 11172-4
  computational accuracy requirements for compliance. In most
  configurations, MAD is a Full Layer III ISO/IEC 11172-3 audio decoder as
  defined by the standard.

  When the approximation version of the fixed-point multiply is used, MAD is
  a limited accuracy ISO/IEC 11172-3 audio decoder as defined by the
  standard.

  MAD can alternatively be configured to produce output with less or more
  accuracy than the default, as a tradeoff with performance.

  MAD produces output samples with a precision greater than 24 bits. Because
  most output formats use fewer bits, typically 16, it is recommended that a
  dithering algorithm be used (rather than rounding or truncating) to obtain
  the highest quality audio. However, dithering may unfavorably affect an
  analytic examination of the output (such as compliance testing); you may
  therefore wish to use rounding in this case instead.

Portability Issues

  GCC is preferred to compile the code, but other compilers may also work.
  The assembly code in `fixed.h' depends on the inline assembly features of
  your compiler. If you're not using GCC or MSVC++, you can either write
  your own assembly macros or use the default (low quality output) version.

  The union initialization of `huffman.c' may not be portable to all
  platforms when GCC is not used.

  The code should not be sensitive to word sizes or byte ordering, however
  it does assume A % B has the same sign as A.

===============================================================================

BUILDING AND INSTALLING

  libmad uses the CMake build system and has no dependencies. To build it, run:

cmake -D CMAKE_INSTALL_PREFIX=/where/you/want/to/install/to -S . -B build
cmake --build build --parallel number-of-cpu-cores
cmake --install build

  The following CMake options are available which can be toggled by passing
  '-D OPTION=VALUE' to the CMake configure step (the first call to cmake above):

  * ASO (ON|OFF): Enable CPU Architecture Specific Optimizations
    (x86, ARM, and MIPS only). ON by default.
  * EXAMPLE (ON|OFF): Build the example application. Only works on POSIX
    operating systems. ON by default on POSIX operating systems. The example
    application is not installed.

===============================================================================

COPYRIGHT

  Please read the `COPYRIGHT' file for copyright and warranty information.
  Also, the file `COPYING' contains the full text of the GNU GPL.

  Send inquiries, comments, bug reports, suggestions, patches, etc. to:

      Underbit Technologies, Inc. <support@underbit.com>

  See also the MAD home page on the Web:

      http://www.underbit.com/products/mad/

===============================================================================



CHANGES

 libmad - MPEG audio decoder library
 Copyright (C) 2000-2004 Underbit Technologies, Inc.

 $Id: CHANGES,v 1.14 2004/02/17 02:02:03 rob Exp $

===============================================================================

Version 0.16.2
  * Fix building assembly file on ARM
  * Fix pkgconfig file when CMAKE_INSTALL_{INCLUDE,LIB}DIR are absolute paths

Version 0.16.1
  * Fix generation of mad.h broken by move to CMake
  * Add CMake options for CPU architecture-specific optimizations

Version 0.16.0
  * Add CMake build system
  * Remove autotools build system
  * Add pkgconfig and CMake config files
  * Apply patches from Debian and Fedora

Version 0.15.1 (beta)

  * Updated to autoconf 2.59, automake 1.8.2, libtool 1.5.2.

  * Replaced Layer III IMDCT routine with one based on a faster algorithm,
    improving both speed and accuracy.

  * Improved portability of the Huffman table initialization.

  * Fixed a problem that could result in an assertion failure in layer3.c
    due to an invalid Layer III free format bitrate.

  * Improved the robustness of Layer II bitrate/mode combinations, and added
    a new MAD_ERROR_BADMODE error enum. The allowability of low-bitrate
    stereo streams is influenced by the --enable-strict-iso option to
    `configure'.

Version 0.15.0 (beta)

  * Updated to autoconf 2.57, automake 1.7.5, libtool 1.4.3.

  * Added new mad_f_div() API routine.

  * Added a 64th entry to the Layer I/Layer II scalefactor table, for better
    compatibility with existing streams. The --enable-strict-iso option to
    `configure' can be used to disable use of this entry.

  * Modified the header decoding routine to allow the reserved emphasis
    value, for better compatibility with existing streams. The
    --enable-strict-iso option to `configure' can be used to restore the
    previous behavior of reporting this value as an error.

  * Added new MAD_EMPHASIS_RESERVED enumeration constant.

  * Fixed a bug in the ARM version of mad_f_scale64() discovered by Andre
    McCurdy.

  * Rewrote PowerPC assembly for minor gains.

  * Modified mad_timer_fraction() to avoid the possibility of division by
    zero when 0 is passed as the second argument.

  * Fixed a non-fatal problem caused by attempting to designate ancillary
    bits in Layer III after a decoding error.

  * Changed to build a shared library by default.

  * Changed to use native Cygwin build by default; give --host=mingw32 to
    `configure' to use MinGW (and avoid a dependency on the Cygwin DLL).

Version 0.14.2 (beta)

  * Changed Cygwin builds to use MinGW; resulting Win32 executables no
    longer have a dependency on Cygwin DLLs.

  * Added a new mad_stream_errorstr() API function to libmad for retrieving
    a string description of the current error condition.

Version 0.14.1 (beta)

  * Updated config.guess and config.sub to latest upstream versions.

  * Enabled libtool versioning rather than release numbering.

  * Improved the documentation in minimad.c.

  * Several other small fixes.

Version 0.14.0 (beta)

  * Added a 64-bit FPM negation operation to improve performance of subband
    synthesis on some platforms.

  * Improved MSVC++ portability and added MSVC++ project files.

  * Added rounding to Layer III requantization for slightly better accuracy.

Version 0.13.0 (beta)

  * Ancillary data is now properly extracted from Layer III streams.

  * Rewrote the Layer III joint stereo decoding routine to correct a major
    MPEG-2 problem and a minor MPEG-1 problem decoding intensity stereo.

  * Eliminated the dependency on sign-extending right shifts for Layer I and
    Layer II.

  * Renamed `private' field to `private_bits' for better C++ compatibility.

  * Gratuitously renamed `sfreq' field to `samplerate' and
    MAD_ERROR_BADSAMPLEFREQ constant to MAD_ERROR_BADSAMPLERATE.

  * Added `samplerate' and `channels' fields to synth.pcm struct to allow
    these to be different from the decoded frame, and for simpler access.

  * Added new mad_stream_options() and mad_decoder_options() API entries for
    special runtime decoding options.

  * Added new MAD_OPTION_IGNORECRC and MAD_OPTION_HALFSAMPLERATE options.

  * Added new MAD_FLAG_FREEFORMAT indicator flag.

  * Fixed some bugs in the async decoder.

  * Added a new mad_timer_multiply() API routine.

  * Eliminated `+' from asm constraints under Intel for better compatibility
    with some compilers.

  * Fixed a PIC-related problem in imdct_l_arm.S.

  * Eliminated a static variable to make libmad thread-safe.

Version 0.12.5 (beta)

  * Modified Layer III requantization to occur during Huffman decoding for
    significant performance gains.

  * Optimized short block IMDCT by eliminating redundant calculations.

  * Made several other Layer III performance improvements; added
    ASO_INTERLEAVE1, ASO_INTERLEAVE2, and ASO_ZEROCHECK
    architecture-specific options for best performance on various
    architectures.

  * Optimized synthesis DCT to store result values as soon as they are
    calculated.

Version 0.12.4 (beta)

  * New PowerPC fixed-point assembly courtesy of David Blythe.

  * Reorganized fixed-point assembly routines for easier maintenance and
    better performance.

  * Improved performance of subband synthesis through better indexing and
    fewer local variables.

  * Added alias reduction for the lower two subbands of mixed short blocks,
    per a report of ambiguity with ISO/IEC 11172-3 and for uniformity with
    most other implementations. Also improved alias reduction performance
    using multiply/accumulate.

  * Added --enable-strict-iso option to `configure' to override best
    accepted practices such as the alias reduction for mixed short blocks.

  * Improved performance of Layer III IMDCT by using longer
    multiply/accumulate runs where possible.

Version 0.12.3 (beta)

  * Added MPEG 2.5 support.

  * Added preliminary support for parameterizing the binary point position

layer12.c
/*
 * libmad - MPEG audio decoder library
 * Copyright (C) 2000-2004 Underbit Technologies, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * $Id: layer12.c,v 1.17 2004/02/05 09:02:39 rob Exp $
 */

# include "global.h"

# ifdef HAVE_LIMITS_H
#  include <limits.h>
# else
#  define CHAR_BIT  8
# endif

# include "fixed.h"
# include "bit.h"
# include "stream.h"
# include "frame.h"
# include "layer12.h"

/*
 * scalefactor table
 * used in both Layer I and Layer II decoding
 */
static
mad_fixed_t const sf_table[64] = {
# include "sf_table.dat"
};

/* --- Layer I ------------------------------------------------------------- */

/* linear scaling table */
static
mad_fixed_t const linear_table[14] = {
  MAD_F(0x15555555),  /* 2^2  / (2^2  - 1) == 1.33333333333333 */
  MAD_F(0x12492492),  /* 2^3  / (2^3  - 1) == 1.14285714285714 */
  MAD_F(0x11111111),  /* 2^4  / (2^4  - 1) == 1.06666666666667 */
  MAD_F(0x10842108),  /* 2^5  / (2^5  - 1) == 1.03225806451613 */
  MAD_F(0x10410410),  /* 2^6  / (2^6  - 1) == 1.01587301587302 */
  MAD_F(0x10204081),  /* 2^7  / (2^7  - 1) == 1.00787401574803 */
  MAD_F(0x10101010),  /* 2^8  / (2^8  - 1) == 1.00392156862745 */
  MAD_F(0x10080402),  /* 2^9  / (2^9  - 1) == 1.00195694716243 */
  MAD_F(0x10040100),  /* 2^10 / (2^10 - 1) == 1.00097751710655 */
  MAD_F(0x10020040),  /* 2^11 / (2^11 - 1) == 1.00048851978505 */
  MAD_F(0x10010010),  /* 2^12 / (2^12 - 1) == 1.00024420024420 */
  MAD_F(0x10008004),  /* 2^13 / (2^13 - 1) == 1.00012208521548 */
  MAD_F(0x10004001),  /* 2^14 / (2^14 - 1) == 1.00006103888177 */
  MAD_F(0x10002000)   /* 2^15 / (2^15 - 1) == 1.00003051850948 */
};

/*
 * NAME:	I_sample()
 * DESCRIPTION:	decode one requantized Layer I sample from a bitstream
 */
static
mad_fixed_t I_sample(struct mad_bitptr *ptr, unsigned int nb, struct mad_stream *stream)
{
  mad_fixed_t sample;
  struct mad_bitptr frameend_ptr;

  mad_bit_init(&frameend_ptr, stream->next_frame);

  if (mad_bit_length(ptr, &frameend_ptr) < nb) {
    stream->error = MAD_ERROR_LOSTSYNC;
    stream->sync = 0;
    return 0;
  }
  sample = mad_bit_read(ptr, nb);

  /* invert most significant bit, extend sign, then scale to fixed format */

  sample ^= 1 << (nb - 1);
  sample |= -(sample & (1 << (nb - 1)));

  sample <<= MAD_F_FRACBITS - (nb - 1);

  /* requantize the sample */

  /* s'' = (2^nb / (2^nb - 1)) * (s''' + 2^(-nb + 1)) */

  sample += MAD_F_ONE >> (nb - 1);

  return mad_f_mul(sample, linear_table[nb - 2]);

  /* s' = factor * s'' */
  /* (to be performed by caller) */
}

/*
 * NAME:	layer->I()
 * DESCRIPTION:	decode a single Layer I frame
 */
int mad_layer_I(struct mad_stream *stream, struct mad_frame *frame)
{
  struct mad_header *header = &frame->header;
  unsigned int nch, bound, ch, s, sb, nb;
  unsigned char allocation[2][32], scalefactor[2][32];
  struct mad_bitptr bufend_ptr, frameend_ptr;

  mad_bit_init(&bufend_ptr, stream->bufend);
  mad_bit_init(&frameend_ptr, stream->next_frame);

  nch = MAD_NCHANNELS(header);

  bound = 32;
  if (header->mode == MAD_MODE_JOINT_STEREO) {
    header->flags |= MAD_FLAG_I_STEREO;
    bound = 4 + header->mode_extension * 4;
  }

  /* check CRC word */

  if (header->flags & MAD_FLAG_PROTECTION) {
    if (mad_bit_length(&stream->ptr, &bufend_ptr)
		< 4 * (bound * nch + (32 - bound))) {
      stream->error = MAD_ERROR_BADCRC;
      return -1;
    }
    header->crc_check =
      mad_bit_crc(stream->ptr, 4 * (bound * nch + (32 - bound)),
		  header->crc_check);

    if (header->crc_check != header->crc_target &&
	!(frame->options & MAD_OPTION_IGNORECRC)) {
      stream->error = MAD_ERROR_BADCRC;
      return -1;
    }
  }

  /* decode bit allocations */

  for (sb = 0; sb < bound; ++sb) {
    for (ch = 0; ch < nch; ++ch) {
      if (mad_bit_length(&stream->ptr, &frameend_ptr) < 4) {
	stream->error = MAD_ERROR_LOSTSYNC;
	stream->sync = 0;
	return -1;
      }
      nb = mad_bit_read(&stream->ptr, 4);

      if (nb == 15) {
	stream->error = MAD_ERROR_BADBITALLOC;
	return -1;
      }

      allocation[ch][sb] = nb ? nb + 1 : 0;
    }
  }

  for (sb = bound; sb < 32; ++sb) {
    if (mad_bit_length(&stream->ptr, &frameend_ptr) < 4) {
      stream->error = MAD_ERROR_LOSTSYNC;
      stream->sync = 0;
      return -1;
    }
    nb = mad_bit_read(&stream->ptr, 4);

    if (nb == 15) {
      stream->error = MAD_ERROR_BADBITALLOC;
      return -1;
    }

    allocation[0][sb] =
    allocation[1][sb] = nb ? nb + 1 : 0;
  }

  /* decode scalefactors */

  for (sb = 0; sb < 32; ++sb) {
    for (ch = 0; ch < nch; ++ch) {
      if (allocation[ch][sb]) {
        if (mad_bit_length(&stream->ptr, &frameend_ptr) < 6) {
	  stream->error = MAD_ERROR_LOSTSYNC;
	  stream->sync = 0;
	  return -1;
	}
	scalefactor[ch][sb] = mad_bit_read(&stream->ptr, 6);

# if defined(OPT_STRICT)
	/*
	 * Scalefactor index 63 does not appear in Table B.1 of
	 * ISO/IEC 11172-3. Nonetheless, other implementations accept it,
	 * so we only reject it if OPT_STRICT is defined.
	 */
	if (scalefactor[ch][sb] == 63) {
	  stream->error = MAD_ERROR_BADSCALEFACTOR;
	  return -1;
	}
# endif
      }
    }
  }

  /* decode samples */

  for (s = 0; s < 12; ++s) {
    for (sb = 0; sb < bound; ++sb) {
      for (ch = 0; ch < nch; ++ch) {
	nb = allocation[ch][sb];
	frame->sbsample[ch][s][sb] = nb ?
	  mad_f_mul(I_sample(&stream->ptr, nb, stream),
		    sf_table[scalefactor[ch][sb]]) : 0;
	if (stream->error != 0)
	  return -1;
      }
    }

    for (sb = bound; sb < 32; ++sb) {
      if ((nb = allocation[0][sb])) {
	mad_fixed_t sample;

	if (mad_bit_length(&stream->ptr, &frameend_ptr) < nb) {
	  stream->error = MAD_ERROR_LOSTSYNC;
	  stream->sync = 0;
          return -1;
	}
	sample = I_sample(&stream->ptr, nb, stream);
        if (stream->error != 0)
	  return -1;

	for (ch = 0; ch < nch; ++ch) {
	  frame->sbsample[ch][s][sb] =
	    mad_f_mul(sample, sf_table[scalefactor[ch][sb]]);
	}
      }
      else {
	for (ch = 0; ch < nch; ++ch)
	  frame->sbsample[ch][s][sb] = 0;
      }
    }
  }

  return 0;
}

/* --- Layer II ------------------------------------------------------------ */

/* possible quantization per subband table */
static
struct {
  unsigned int sblimit;
  unsigned char const offsets[30];
} const sbquant_table[5] = {
  /* ISO/IEC 11172-3 Table B.2a */
  { 27, { 7, 7, 7, 6, 6, 6, 6, 6, 6, 6, 6, 3, 3, 3, 3, 3,	/* 0 */
	  3, 3, 3, 3, 3, 3, 3, 0, 0, 0, 0 } },
  /* ISO/IEC 11172-3 Table B.2b */
  { 30, { 7, 7, 7, 6, 6, 6, 6, 6, 6, 6, 6, 3, 3, 3, 3, 3,	/* 1 */
	  3, 3, 3, 3, 3, 3, 3, 0, 0, 0, 0, 0, 0, 0 } },
  /* ISO/IEC 11172-3 Table B.2c */
  {  8, { 5, 5, 2, 2, 2, 2, 2, 2 } },				/* 2 */
  /* ISO/IEC 11172-3 Table B.2d */
  { 12, { 5, 5, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2 } },		/* 3 */
  /* ISO/IEC 13818-3 Table B.1 */
  { 30, { 4, 4, 4, 4, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1,	/* 4 */
	  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 } }
};

/* bit allocation table */
static
struct {
  unsigned short nbal;
  unsigned short offset;
} const bitalloc_table[8] = {
  { 2, 0 },  /* 0 */
  { 2, 3 },  /* 1 */
  { 3, 3 },  /* 2 */
  { 3, 1 },  /* 3 */
  { 4, 2 },  /* 4 */
  { 4, 3 },  /* 5 */
  { 4, 4 },  /* 6 */
  { 4, 5 }   /* 7 */
};

/* offsets into quantization class table */
static
unsigned char const offset_table[6][15] = {
  { 0, 1, 16                                             },  /* 0 */
  { 0, 1,  2, 3, 4, 5, 16                                },  /* 1 */
  { 0, 1,  2, 3, 4, 5,  6, 7,  8,  9, 10, 11, 12, 13, 14 },  /* 2 */
  { 0, 1,  3, 4, 5, 6,  7, 8,  9, 10, 11, 12, 13, 14, 15 },  /* 3 */
  { 0, 1,  2, 3, 4, 5,  6, 7,  8,  9, 10, 11, 12, 13, 16 },  /* 4 */
  { 0, 2,  4, 5, 6, 7,  8, 9, 10, 11, 12, 13, 14, 15, 16 }   /* 5 */
};

/* quantization class table */
static
struct quantclass {
  unsigned short nlevels;
  unsigned char group;
  unsigned char bits;
  mad_fixed_t C;
  mad_fixed_t D;
} const qc_table[17] = {
# include "qc_table.dat"
};

/*
 * NAME:	II_samples()
 * DESCRIPTION:	decode three requantized Layer II samples from a bitstream
 */
static
void II_samples(struct mad_bitptr *ptr,
		struct quantclass const *quantclass,
		mad_fixed_t output[3], struct mad_stream *stream)
{
  unsigned int nb, s, sample[3];
  struct mad_bitptr frameend_ptr;

  mad_bit_init(&frameend_ptr, stream->next_frame);

  if ((nb = quantclass->group)) {
    unsigned int c, nlevels;

    if (mad_bit_length(ptr, &frameend_ptr) < quantclass->bits) {
      stream->error = MAD_ERROR_LOSTSYNC;
      stream->sync = 0;
      return;
    }
    /* degrouping */
    c = mad_bit_read(ptr, quantclass->bits);
    nlevels = quantclass->nlevels;

    for (s = 0; s < 3; ++s) {
      sample[s] = c % nlevels;
      c /= nlevels;
    }
  }
  else {
    nb = quantclass->bits;

    for (s = 0; s < 3; ++s) {
      if (mad_bit_length(ptr, &frameend_ptr) < nb) {
	stream->error = MAD_ERROR_LOSTSYNC;
	stream->sync = 0;
	return;
      }
      sample[s] = mad_bit_read(ptr, nb);
    }
  }

  for (s = 0; s < 3; ++s) {
    mad_fixed_t requantized;

    /* invert most significant bit, extend sign, then scale to fixed format */

    requantized  = sample[s] ^ (1 << (nb - 1));
    requantized |= -(requantized & (1 << (nb - 1)));

    requantized <<= MAD_F_FRACBITS - (nb - 1);

    /* requantize the sample */

    /* s'' = C * (s''' + D) */

    output[s] = mad_f_mul(requantized + quantclass->D, quantclass->C);

    /* s' = factor * s'' */
    /* (to be performed by caller) */
  }
}

/*
 * NAME:	layer->II()
 * DESCRIPTION:	decode a single Layer II frame
 */
int mad_layer_II(struct mad_stream *stream, struct mad_frame *frame)
{
  struct mad_header *header = &frame->header;
  struct mad_bitptr start;
  unsigned int index, sblimit, nbal, nch, bound, gr, ch, s, sb;
  unsigned char const *offsets;
  unsigned char allocation[2][32], scfsi[2][32], scalefactor[2][32][3];
  mad_fixed_t samples[3];
  struct mad_bitptr frameend_ptr;

  mad_bit_init(&frameend_ptr, stream->next_frame);

  nch = MAD_NCHANNELS(header);

  if (header->flags & MAD_FLAG_LSF_EXT)
    index = 4;
  else if (header->flags & MAD_FLAG_FREEFORMAT)
    goto freeformat;
  else {
    unsigned long bitrate_per_channel;

    bitrate_per_channel = header->bitrate;
    if (nch == 2) {
      bitrate_per_channel /= 2;

# if defined(OPT_STRICT)
      /*
       * ISO/IEC 11172-3 allows only single channel mode for 32, 48, 56, and
       * 80 kbps bitrates in Layer II, but some encoders ignore this
       * restriction. We enforce it if OPT_STRICT is defined.
       */
      if (bitrate_per_channel <= 28000 || bitrate_per_channel == 40000) {
	stream->error = MAD_ERROR_BADMODE;
	return -1;
      }
# endif
    }
    else {  /* nch == 1 */
      if (bitrate_per_channel > 192000) {
	/*
	 * ISO/IEC 11172-3 does not allow single channel mode for 224, 256,
	 * 320, or 384 kbps bitrates in Layer II.
	 */
	stream->error = MAD_ERROR_BADMODE;
	return -1;
      }
    }

    if (bitrate_per_channel <= 48000)
      index = (header->samplerate == 32000) ? 3 : 2;
    else if (bitrate_per_channel <= 80000)
      index = 0;
    else {
    freeformat:
      index = (header->samplerate == 48000) ? 0 : 1;
    }
  }

  sblimit = sbquant_table[index].sblimit;
  offsets = sbquant_table[index].offsets;

  bound = 32;
  if (header->mode == MAD_MODE_JOINT_STEREO) {
    header->flags |= MAD_FLAG_I_STEREO;
    bound = 4 + header->mode_extension * 4;
  }

  if (bound > sblimit)
    bound = sblimit;

  start = stream->ptr;

  /* decode bit allocations */

  for (sb = 0; sb < bound; ++sb) {
    nbal = bitalloc_table[offsets[sb]].nbal;

    for (ch = 0; ch < nch; ++ch) {
      if (mad_bit_length(&stream->ptr, &frameend_ptr) < nbal) {
	stream->error = MAD_ERROR_LOSTSYNC;
	stream->sync = 0;
	return -1;
      }
      allocation[ch][sb] = mad_bit_read(&stream->ptr, nbal);
    }
  }

  for (sb = bound; sb < sblimit; ++sb) {
    nbal = bitalloc_table[offsets[sb]].nbal;

    if (mad_bit_length(&stream->ptr, &frameend_ptr) < nbal) {
      stream->error = MAD_ERROR_LOSTSYNC;
      stream->sync = 0;
      return -1;
    }
    allocation[0][sb] =
    allocation[1][sb] = mad_bit_read(&stream->ptr, nbal);
  }

  /* decode scalefactor selection info */

  for (sb = 0; sb < sblimit; ++sb) {
    for (ch = 0; ch < nch; ++ch) {
      if (allocation[ch][sb]) {
	if (mad_bit_length(&stream->ptr, &frameend_ptr) < 2) {
	  stream->error = MAD_ERROR_LOSTSYNC;
	  stream->sync = 0;
	  return -1;
	}
	scfsi[ch][sb] = mad_bit_read(&stream->ptr, 2);
      }
    }
  }

  /* check CRC word */

  if (header->flags & MAD_FLAG_PROTECTION) {
    header->crc_check =
      mad_bit_crc(start, mad_bit_length(&start, &stream->ptr),
		  header->crc_check);

    if (header->crc_check != header->crc_target &&
	!(frame->options & MAD_OPTION_IGNORECRC)) {
      stream->error = MAD_ERROR_BADCRC;
      return -1;
    }
  }

  /* decode scalefactors */

  for (sb = 0; sb < sblimit; ++sb) {
    for (ch = 0; ch < nch; ++ch) {
      if (allocation[ch][sb]) {
	if (mad_bit_length(&stream->ptr, &frameend_ptr) < 6) {
	  stream->error = MAD_ERROR_LOSTSYNC;
	  stream->sync = 0;
	  return -1;
	}
	scalefactor[ch][sb][0] = mad_bit_read(&stream->ptr, 6);

	switch (scfsi[ch][sb]) {
	case 2:
	  scalefactor[ch][sb][2] =
	  scalefactor[ch][sb][1] =
	  scalefactor[ch][sb][0];
	  break;

	case 0:
	  if (mad_bit_length(&stream->ptr, &frameend_ptr) < 6) {
	    stream->error = MAD_ERROR_LOSTSYNC;
	    stream->sync = 0;
	    return -1;
	  }
	  scalefactor[ch][sb][1] = mad_bit_read(&stream->ptr, 6);
	  /* fall through */

	case 1:
	case 3:
	  if (mad_bit_length(&stream->ptr, &frameend_ptr) < 6) {
	    stream->error = MAD_ERROR_LOSTSYNC;
	    stream->sync = 0;
	    return -1;
	  }
	  scalefactor[ch][sb][2] = mad_bit_read(&stream->ptr, 6);
	}

	if (scfsi[ch][sb] & 1)
	  scalefactor[ch][sb][1] = scalefactor[ch][sb][scfsi[ch][sb] - 1];

# if defined(OPT_STRICT)
	/*
	 * Scalefactor index 63 does not appear in Table B.1 of
	 * ISO/IEC 11172-3. Nonetheless, other implementations accept it,
	 * so we only reject it if OPT_STRICT is defined.
	 */
	if (scalefactor[ch][sb][0] == 63 ||
	    scalefactor[ch][sb][1] == 63 ||
	    scalefactor[ch][sb][2] == 63) {
	  stream->error = MAD_ERROR_BADSCALEFACTOR;
	  return -1;
	}
# endif
      }
    }
  }

  /* decode samples */

  for (gr = 0; gr < 12; ++gr) {
    for (sb = 0; sb < bound; ++sb) {
      for (ch = 0; ch < nch; ++ch) {
	if ((index = allocation[ch][sb])) {
	  index = offset_table[bitalloc_table[offsets[sb]].offset][index - 1];

	  II_samples(&stream->ptr, &qc_table[index], samples, stream);
	  if (stream->error != 0)
            return -1;

	  for (s = 0; s < 3; ++s) {
	    frame->sbsample[ch][3 * gr + s][sb] =
	      mad_f_mul(samples[s], sf_table[scalefactor[ch][sb][gr / 4]]);
	  }
	}
	else {
	  for (s = 0; s < 3; ++s)
	    frame->sbsample[ch][3 * gr + s][sb] = 0;
	}
      }
    }

    for (sb = bound; sb < sblimit; ++sb) {
      if ((index = allocation[0][sb])) {
	index = offset_table[bitalloc_table[offsets[sb]].offset][index - 1];

	II_samples(&stream->ptr, &qc_table[index], samples, stream);
	if (stream->error != 0)
          return -1;

	for (ch = 0; ch < nch; ++ch) {
	  for (s = 0; s < 3; ++s) {
	    frame->sbsample[ch][3 * gr + s][sb] =
	      mad_f_mul(samples[s], sf_table[scalefactor[ch][sb][gr / 4]]);
	  }
	}
      }
      else {
	for (ch = 0; ch < nch; ++ch) {
	  for (s = 0; s < 3; ++s)
	    frame->sbsample[ch][3 * gr + s][sb] = 0;
	}
      }
    }

    for (ch = 0; ch < nch; ++ch) {
      for (s = 0; s < 3; ++s) {
	for (sb = sblimit; sb < 32; ++sb)
	  frame->sbsample[ch][3 * gr + s][sb] = 0;
      }
    }
  }

  return 0;
}
